drivers: flash controller for the stm32u5 soc serie

FRASTM opened the pull request:
This PR adds the flash controller driver for the stm32U5xx soc
It defines the flash controller node in the DTS
It defines a partition for the flash memory. 
The drivers/flash/flash_stm32u5x.c  is adapted from the drivers/flash/flash_stm32l5x.c
Applicable on the stm32U575/585 devices where the Flash memory is organized as two banks of 1 Mbyte each containing 128 pages of 8 Kbytes ( 137 bits wide data read and write = 128 effective bits plus 9 ECC bits)

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/42478

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: flash for the stm32l5x or stm32u5 serie with max 2MB

Adds the stm32u5 flash controller driver for this serie
to the existing stm32l5 flash driver part
Only 1 or 2 MB devices exist today (4MB is possible in the future).
This flash controller driver is adapted from the flash_stm32l5.c

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
boards: arm: stm32 u585 disco kit set flash partition

Enables the flash controller node for the stm32u585 disco kit
with an arbitrary definition of the board partitions
to fill the 1MB bank1 of the flash memory,
for the secure and non-secure targets.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: flash stm32 flash driver includes stm32u5 serie

This adds the stm32U5 soc family to the flash driver
The flash controller has particular register names in the Non-Secure
area to be adapted for the driver.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
dts: arm: stm32u5 family with flash controller node

This commit adds a flash_controller node for the stm32u5 soc family
blocks of 8K and erase time given by the Datasheet (with margin).

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
$ west build -p auto -b b_u585i_iot02a tests/drivers/flash
```
*** Booting Zephyr OS build v3.0.0-rc1-252-g7de53a4c47a2  ***                   
Running test suite flash_driver_test                                            
===================================================================             
START - test_setup                                                              
 PASS - test_setup in 0.1 seconds                                               
===================================================================             
START - test_read_unaligned_address                                             
 PASS - test_read_unaligned_address in 0.4 seconds                              
===================================================================             
Test suite flash_driver_test succeeded                                          
===================================================================             
PROJECT EXECUTION SUCCESSFUL 
```

FRASTM commented:
adding the DTS for the **b_u585i_iot02a_ns** target  with a `image-1-nonsecure` partition

FRASTM commented:
rebase

wcappelle commented on drivers/flash/flash_stm32u5x.c:
```diff
@@ -0,0 +1,380 @@
+	}
+
+	/* Check if this double word is erased */
+	if ((flash[0] != 0xFFFFFFFFUL) || (flash[1] != 0xFFFFFFFFUL)) {
```
just a note: a write will still fail if the ECC is already written.
ECC can be set if a programmer/ previous write was done with 0xFF...FF as data.
Maybe good to add this in the check?

erwango commented on drivers/flash/flash_stm32u5x.c:
```diff
@@ -0,0 +1,380 @@
+#define ICACHE_DISABLE_TIMEOUT_VALUE           1U   /* 1ms */
+#define ICACHE_INVALIDATE_TIMEOUT_VALUE        1U   /* 1ms */
+
+static int stm32u5x_icache_disable(void)
```
I wonder what is the diff vs L5 driver.
If close enough, it would be good to use a single file to ease maintainance.
For instance https://github.com/zephyrproject-rtos/zephyr/pull/42099 may need to be backported on this file also is duplicated

FRASTM commented on drivers/flash/flash_stm32l5_u5.c:
```diff
@@ -0,0 +1,396 @@
+		 * or For stm32u58x with 2MB flash
+		 */
+
+		if (stm32_flash_layout[0].pages_count == 0) {
```
@wcappelle could you please detail what you mean, I do not catch your comment

erwango commented on drivers/flash/flash_stm32.h:
```diff
@@ -53,6 +53,19 @@ struct flash_stm32_priv {
 
 
 /* Redefinitions of flags and masks to harmonize stm32 series: */
+#if defined(CONFIG_SOC_SERIES_STM32U5X)
+#define FLASH_NSCR_NSLOCK FLASH_NSCR_LOCK
+#define FLASH_OPTR_DBANK FLASH_OPTR_DUALBANK
+#define FLASH_NSCR_NSPG FLASH_NSCR_PG
+#define FLASH_NSCR_NSBKER_Msk FLASH_NSCR_BKER_Msk
+#define FLASH_NSCR_NSBKER FLASH_NSCR_BKER
+#define FLASH_NSCR_NSPER FLASH_NSCR_PER
+#define FLASH_NSCR_NSPNB_Msk FLASH_NSCR_PNB_Msk
+#define FLASH_NSCR_NSPNB_Pos FLASH_NSCR_PNB_Pos
+#define FLASH_NSCR_NSPNB FLASH_NSCR_PNB
+#define FLASH_NSCR_NSSTRT FLASH_NSCR_STRT
+#endif /* CONFIG_SOC_SERIES_STM32U5X */
+
```
For clarity, if we need to provide zephyr specific abstraction on top of Cube apis, avoid re-using Cube definition and create zephyr specific ones.

FRASTM replied:
the fact is that FLASH_NSCR_NSLOCK is in the flash_stm32.c or FLASH_NSCR_NSPNB_Msk in the flash_stm32l5x.c etc

erwango replied:
Yes

FRASTM replied:
The FLASH_NSCR_LOCK is the name of the bit registered in the stm32CUbe/stm32l5x HAL
Changing FLASH_NSCR_LOCK to FLASH_STM32_NSCR_LOCK in the zephyr driver means 
  a change in the drivers/flash/stm32_flash.h and also updating drivers/flash/flash_stm32l5x.c, drivers/flash/flash_stm32g4x.c and drivers/flash/flash_stm32l4x

FRASTM replied:
cf commit 4a94eedbc1a428f1013a662a318022a6d5499eb9

erwango replied:
I agree this is not straightforward but we should keep each level of abstraction clear. Otherwise this will lead to confusion in the end (if this is spreading in other drivers)
Alternatives are:
- get CMSIS definitions fixed (but LL is here to provide CMSIS abstraction)
- convert whole driver to use of LL

erwango commented on dts/arm/st/u5/stm32u5.dtsi:
```diff
@@ -83,8 +84,8 @@
-		flash-controller@40022000 {
-			compatible = "st,stm32u5-flash-controller";
+		flash: flash-controller@40022000 {
+			compatible = "st,stm32-flash-controller", "st,stm32wb-flash-controller";
```
stm32wb ?

erwango commented on drivers/flash/flash_stm32l5_u5.c:
```diff
@@ -0,0 +1,405 @@
+/*
```
I wonder why this is appearing as a new file introduction.
Maybe you need to remove the L5 one ?

FRASTM commented:
rebase on ae71444426a07a918ff82132e046e90d05093725
deleting flash_stm32l5.c

erwango commented on boards/arm/b_u585i_iot02a/b_u585i_iot02a.dts:
```diff
@@ -24,3 +24,38 @@
 };
+
+&flash0 {
+	/*
+	 * For more information, see:
+	 * https://docs.zephyrproject.org/latest/reference/devicetree/index.html#fixed-flash-partitions
+	 */
```
Please remove this

erwango replied:
cf https://github.com/zephyrproject-rtos/zephyr/pull/43882

FRASTM replied:
removed

FRASTM committed:
drivers: flash: stm32 definition of all registered bitfield

redefine all the stm32 flash register bit Name from
FLASH_NSCR_xxx to FLASH_STM32_FLASH_NSCR_xxx
in all the zephyr drivers.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
add a commit 4a94eedbc1a428f1013a662a318022a6d5499eb9 to change FLASH_STM32_xx bitfield

FRASTM commented:
rebase on 574e166e1357641cb80dda00ff0a01096b6695a0
reorder commits
fix compilation error on some l4 boards

erwango commented on drivers/flash/Kconfig.stm32:
```diff
@@ -7,7 +7,7 @@
 	bool "STM32 flash driver"
 	depends on SOC_FAMILY_STM32
-	depends on (SOC_SERIES_STM32F0X || SOC_SERIES_STM32F1X || SOC_SERIES_STM32F2X || SOC_SERIES_STM32F3X || SOC_SERIES_STM32F4X || SOC_SERIES_STM32F7X || SOC_SERIES_STM32L0X  || SOC_SERIES_STM32L1X || SOC_SERIES_STM32L4X || SOC_SERIES_STM32L5X || SOC_SERIES_STM32WBX || SOC_SERIES_STM32WLX || SOC_SERIES_STM32G0X || SOC_SERIES_STM32G4X || SOC_SERIES_STM32H7X)
+	depends on (SOC_SERIES_STM32F0X || SOC_SERIES_STM32F1X || SOC_SERIES_STM32F2X || SOC_SERIES_STM32F3X || SOC_SERIES_STM32F4X || SOC_SERIES_STM32F7X || SOC_SERIES_STM32L0X  || SOC_SERIES_STM32L1X || SOC_SERIES_STM32L4X || SOC_SERIES_STM32L5X || SOC_SERIES_STM32U5X || SOC_SERIES_STM32WBX || SOC_SERIES_STM32WLX || SOC_SERIES_STM32G0X || SOC_SERIES_STM32G4X || SOC_SERIES_STM32H7X)
```
Not blocking: would be nice to replace this line by a :
```
DT_COMPAT_ST_STM32_FLASH_CONTROLLER := st,stm32-flash-controller
[...]
	depends on $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_FLASH_CONTROLLER))
```

erwango replied:
Note: This might need some tests. I'm not sure how this will behave when there is no flash controller available for the target.

FRASTM replied:
ok

FRASTM replied:
with a special case for H7 : `DT_COMPAT_ST_STM32H7_FLASH_CONTROLLER := st,stm32h7-flash-controller`

erwango commented on drivers/flash/flash_stm32l5_u5.c:
```diff
@@ -15,18 +15,30 @@ LOG_MODULE_REGISTER(LOG_DOMAIN);
 #include <drivers/flash.h>
 #include <init.h>
 #include <soc.h>
+#if defined(CONFIG_SOC_SERIES_STM32L5X)
 #include <stm32l5xx_ll_icache.h>
+#endif
+#if defined(CONFIG_SOC_SERIES_STM32U5X)
+#include <stm32u5xx_ll_icache.h>
+#endif
+
```
Could be replaced by #include <stm32_ll_icache.h>

erwango commented on boards/arm/b_u585i_iot02a/b_u585i_iot02a.dts:
```diff
@@ -24,3 +24,34 @@
+
+&flash0 {
+	partitions {
+		compatible = "fixed-partitions";
```
Let's add a comment specifying that this partition is dedicated to the use of b_u585i_iot02a target with TZEN=0 (so w/o TFM)

FRASTM replied:
ok

erwango commented on tests/drivers/flash/boards/b_u585i_iot02a.overlay:
```diff
@@ -0,0 +1,11 @@
+ */
+
+/ {
+	chosen {
+		zephyr,code-partition = &slot0_partition;
+	};
```
Why not putting this directly in board.dts ?

FRASTM commented:
with optimized Kconfig writing and dts

erwango commented on drivers/flash/flash_stm32.h:
```diff
@@ -54,18 +54,31 @@ struct flash_stm32_priv {
-#define FLASH_NSCR_NSLOCK FLASH_NSCR_LOCK
-#define FLASH_OPTR_DBANK FLASH_OPTR_DUALBANK
-#define FLASH_NSCR_NSPG FLASH_NSCR_PG
-#define FLASH_NSCR_NSBKER_Msk FLASH_NSCR_BKER_Msk
```
This whole part should ideally be moved into initial commit "drivers: flash stm32 flash driver includes stm32u5 serie".
But I agree this require a full rework of the PR, so ok to keep it that way.

mbolivar-nordic commented:
Merging since @nvlsianpu is on leave
